btree/btree.c

```diff
@@ -3740,6 +3767,58 @@ int64_t btree_open_tick(struct m0_sm_op *smop)
 	}
 }
 
+/**
+ * btree_close_tree_tick function is used to traverse through different states
+ * to facilitate the working of m0_btree_close().
+ *
+ * @param smop     represents the state machine operation
+ * @return int64_t returns the next state to be executed.
+ */
+int64_t btree_close_tree_tick(struct m0_sm_op *smop)
+{
+	struct m0_btree_op    *bop     = M0_AMB(bop, smop, bo_op);
+	struct m0_btree       *tree    = bop->bo_arbor;
+	struct td             *td_curr = tree->t_desc;
+	struct nd             *nd_head = ndlist_tlist_head(&td_curr->
+							   t_active_nds);
+	struct nd             *nd_curr;
+
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+		if (td_curr->t_ref > 1)
+			tree_put(td_curr);
+		else if (td_curr->t_ref == 1) {
+			nd_curr = nd_head;
+			if (td_curr->t_starttime == 0)
+				td_curr->t_starttime = m0_time_now();
+
+			if (ndlist_tlist_length(&td_curr->t_active_nds) > 1)
+				return P_TIMECHECK;
+			else if (nd_curr == td_curr->t_root)
+				node_put(nd_curr);
+
+			td_curr->t_starttime = 0;
+			tree_put(td_curr);
+		} else
+			return M0_ERR(-ECANCELED);
```

Comment:
This is invalid scenario, there should be an assert at the beginning for t_ref = 0.

Reply:
Updated.